docs/discussion/Cosign Upgrade 2024.md

```diff
@@ -0,0 +1,457 @@
+
+> **NOTE**: Based on Cosign's own verification implementation, there does not seem to be a way to enforce that ALL signatures found are valid. This behavior is unchartered as far as we can tell and would require Ratify to determine if we want to support this directly. For that reason, the strict ALL signature scenario is not yet included.
+
+#### Extra Considerations
+- How does Kyverno handle multiple Keys?
+  - Each key specified is a separate cosign verification operation
+  - User specifies enforcement type by specifying `count` of verification operations that must succeed.
+    - e.g 2 keys specified and 2 signatures on the image. each verification is against both signatures for a single key
+    - If count is 2, then 2 verifications must return true
+- Why is `keyEnforcement` necessary?
+  - Not necessary if multiple cosign verifiers are used, each configured with different keys.
+  - `keyEnforcement` allows for single verifier to be used instead of multiple
+  - more user friendly
+  - does not require extra policy logic (although this logic will likely be needed anyways once OCI 1.1 is the default standard cosign uses)
+- How do you specify specific AKV keys you trust rather than the whole truststore?
+- How will this work with OCI 1.1 enabled cosign?
+  - Cosign OCI 1.1 support is experimental
+  - Currently, an arbitrary signature from list of referrers is chosen to perform verifications on IF multiple cosign artifacts are found.
+  - This logic will be updated in the future once 1.1 is GA but for now it's left as a TODO in the Cosign code base.
+
+### Implementation Details
+
+#### Handling multiple verifications at once
+- Ratify will largely mirror how cosign performs multi-signature verification
+  - cannot invoke cosign's multi signature verification API since the API exposed also includes the registry content fetch which is independent in Ratify
+- separate go routines per signature and collect all results PER KEY 
+
+## Support RSA and ED25519 keys for verification
+
+Each key specified will encapsulate a single verification context for all signatures found in the artifact that the verifier is running for.
+
+Verification flow per key:
+- Take raw key content (byte array) and [`LoadPublicKeyRaw()`](https://github.com/sigstore/cosign/blob/daf1eeb5ff2022d11466b8eccda25768b2dd2992/pkg/signature/keys.go#L91C1-L91C1)
+- The `LoadPublicKeyRaw()` method in the cosign keys library will automatically detect which key type (based on the key header) and return the appropriate signature verifier
+- Verifier is then passed to cosign's `VerifySignature()` method as a verification option
+
+## Dev Work Items (WIP)
+- Add `CertificateStore` support to CLI
+- Rename `CertificateStore` CRD to `KeyStore` CRD (maintain backwards compatability)
```

Comment:
I did no see why we need to rename `CertificateStore` CRD to `KeyStore` CRD

Reply:
Discussed a bit in community call on 1/18/24. If we determine that the existing CertificateStore name is too confusing to encompass key functionality, we will need to deprecate CertificateStore and introduce a newly named CRD in its place. Something like `KeyStore`. @yizha1 could you help with determining a good name to call it?

Reply:
I created an issue #1287 and ask for votes